test/include/seqan3/test/sandboxed_path.hpp

```diff
@@ -0,0 +1,562 @@
+// Copyright (c) 2006-2021, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2021, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides seqan3::test::sandboxed_path and related free functions.
+ * \author Simon Gene Gottlieb <simon.gottlieb AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <seqan3/std/algorithm>
+#include <seqan3/std/filesystem>
+#include <numeric>
+
+#include <seqan3/core/platform.hpp>
+
+namespace seqan3::test
+{
+
+/*!\brief Utility class to stay inside a sandbox path.
+ *
+ * seqan3::sandboxed_path provides the same functionality as std::filesystem::path, but restricts
+ * the access to a specified directory. This results in the following invariant, which is checked
+ * for at appropiate places, and some caveats.
+ *
+ *  Invariant:
+ *  - seqan3::sandboxed_path is always converted to an absolute path
+ *  - seqan3::sandboxed_path always points to a file or directory inside a given sandbox directory
+ *  - The sandbox directory is immutable during the life cycle of a sandboxed_path
+ *
+ * Caveat:
+ *  - Relative paths are not possible
+ *  - Some functions will leave the sandboxed environment.
+ *    - calling relative_path() leaves the environment of sandboxed_path.
+ */
+class sandboxed_path : public std::filesystem::path
+{
+private:
+    std::filesystem::path const sandbox_directory;
+
+public:
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+
+    /*!\brief Construction of a sandboxed_path.
+     * \param directory must be an absolute path.
+     *
+     * After construction, the sandboxed_path will point to `path`.
+     */
+    explicit sandboxed_path(std::filesystem::path directory)
+        : std::filesystem::path{directory}
+        , sandbox_directory{std::move(directory)}
+    {
+        normalise();
+        check_invariant();
+    }
+
+    /*!\brief Construction from a given sandbox directory and a path within the sandbox directory.
+     * \param sandbox_directory The absolute path to the sandbox directory.
+     * \param path The relative or absolute path that must be inside the sandbox directory.
+     */
+    explicit sandboxed_path(std::filesystem::path sandbox_directory, std::filesystem::path path)
+        : std::filesystem::path {std::move(path)}
+        , sandbox_directory     {std::move(sandbox_directory)}
+    {
+        normalise();
+        check_invariant();
+    }
+
+    sandboxed_path() = delete;                           //!< Deleted.
+    sandboxed_path(sandboxed_path const&) = default;     //!< Defaulted.
+    sandboxed_path(sandboxed_path&&) noexcept = default; //!< Defaulted.
```

Comment:
Copy and move assignment missing

Reply:
They are directly below the constructors, maybe not obviously but they are covered by the template copy/move assignments.
```
    template <typename path_t>
    sandboxed_path & operator=(path_t const & new_path)
...
```

Reply:
We still add the non-template versions.

Reply:
Ah, ok, than I will add them :-)